sled-agent/src/hardware/illumos/mod.rs

```diff
@@ -73,56 +167,194 @@ fn node_name(subsystem_vid: i32, subsystem_id: i32) -> String {
+                // Devices in "/dev/dsk" have names that denote their purpose,
+                // of the form "controller, disk, slice" or "controller, disk,
+                // partition".
+                //
+                // The suffix of "d0" is typical of an individual disk, and is
+                // the expected device to correspond with the "wd" device in
+                // the "/devices" hierarchy.
+                l.linktype() == DevLinkType::Primary
+                    && l.path()
+                        .file_name()
+                        .map(|f| f.to_string_lossy().ends_with("d0"))
+                        .unwrap_or(false)
+            })
+            .collect::<Vec<_>>();
+
+        if paths.is_empty() {
+            return Err(Error::NoDevLinks(PathBuf::from(devfs_path)));
+        }
+        return Ok(Some(paths[0].path().to_path_buf()));
+    }
+    Ok(None)
+}
+
+// Gather a notion of "device identity".
+//
+// NOTE: What exactly comprises a "reliable, unique identifier" for a disk is
+// still a little bit in flux. While that's being sorted out, we simply produce
+// a string that can uniquely identify the device to Nexus.
+//
+// At the moment, this consists of a "Vendor ID, Serial ID, and model ID".
+fn blkdev_device_identity(devfs_path: &str) -> Result<String, Error> {
+    let path = PathBuf::from(devfs_path);
+    let ctrl_id = nvme::identify_controller(&path)
+        .map_err(|err| Error::MissingNVMEIdentity { path, err })?;
+
+    let vendor =
+        format!("{:x}{:x}", ctrl_id.vendor_id, ctrl_id.subsystem_vendor_id);
+    let serial = ctrl_id.serial;
+    let model = ctrl_id.model;
+
+    Ok(format!("v:{vendor},s:{serial},m:{model}"))
+}
+
+fn poll_blkdev_node(
+    log: &Logger,
+    disks: &mut HashSet<UnparsedDisk>,
+    mut node: Node<'_>,
+) -> Result<(), Error> {
+    if let Some(driver_name) = node.driver_name() {
+        if driver_name == "blkdev" {
+            let devfs_path = node.devfs_path().map_err(Error::DevInfo)?;
+            let dev_path = get_dev_path_of_whole_disk(&node)?;
+
+            // For some reason, libdevfs doesn't prepend "/devices" when
+            // referring to the path, but it still returns an absolute path.
+            //
+            // Validate that we're still using this leading slash, but make the
+            // path "real".
+            assert!(devfs_path.starts_with('/'));
+            let devfs_path = format!("/devices{devfs_path}");
+            let mut device_id = None;
+            while let Some(parent) = node.parent().map_err(Error::DevInfo)? {
+                node = parent;
+
+                if let Some(driver_name) = node.driver_name() {
+                    if driver_name == "nvme" {
+                        device_id = Some(blkdev_device_identity(&format!(
+                            "/devices{}:devctl",
+                            node.devfs_path().map_err(Error::DevInfo)?
+                        ))?);
+                    }
+                }
```

Comment:
To make this a bit more concrete:

We might find a device like:
`blkdev` -> `/devices/pci@ab,0/pci1022,1483@1,1/pci1b96,0@0/blkdev@w0014EE81000BC153,0`

This means we could find a parent `nvme` device at:
`nvme` -> `/devices/pci@ab,0/pci1022,1483@1,1/pci1b96,0@0:devctl`

This is the device that will respond to NVMe-specific ioctls, and can return a unique device identity for us to use as a primary key at a higher layer in the control plane.